src/graph/GraphViewServer.ts

```diff
@@ -0,0 +1,168 @@
+import { GraphConfiguration } from './GraphConfiguration';
+import * as gremlin from "gremlin";
+
+/**
+ * @class GraphViewServer This is the server side of the graph explorer. It handles all communications
+ * with Azure including gremlin queries. It communicates with the client code via an HTTP server and
+ * sockets.
+ */
+export class GraphViewServer extends EventEmitter {
+  private _server: SocketIO.Server;
+  private _httpServer: http.Server;
+  private _port: number | undefined;
+  private _socket: SocketIO.Socket;
+
+  constructor(private _configuration: GraphConfiguration) {
+    super();
+  }
+
+  public get configuration(): GraphConfiguration {
+    return this._configuration;
+  }
+
+  public dispose() {
+    if (this._socket) {
+      this._socket.disconnect();
+      this._socket = null;
+    }
+    if (this._httpServer) {
+      this._httpServer.close();
+      this._httpServer = null;
+    }
+    if (this._server) {
+      this._server.close();
+      this._server = null;
+    }
+  }
+
+  public get port(): number {
+    if (!this._port) {
+      throw new Error("Server has not started");
+    }
+
+    return this._port;
+  }
+
+  public start(): Promise<void> {
+    if (this._socket) {
+      return Promise.resolve();
+    }
+
+    // how know resolve/reject?
+    return new Promise((resolve, reject) => {
+      this._httpServer = http.createServer()
+      this._httpServer.listen(
+        0, // dynamnically pick an unused port
+        () => {
+          this._port = this._httpServer.address().port;
+          console.log(`** Server using port ${this._port} for ${this._configuration.endpoint}`);
+          resolve();
+        });
+      this._server = io(this._httpServer);
+
+      this._server.on('connection', socket => { // TODO called multiple times?  what if already a socket?    dispose?
+        console.log(`Connected to client ${socket.id}`);
+        this._socket = socket;
+        this.setUpSocket();
+      });
+
+      this._server.on('error', socket => {
+        console.log("error"); // TODO
+        reject("TODO");
+      });
+    });
+  }
+
+  private async queryAndShowResults(queryId: number, gremlinQuery: string): Promise<void> {
+    var results: any[];
+
+    try {
+      var vertices = await this.executeQuery(queryId, gremlinQuery);
+      results = vertices;
+
+      // If it returned any vertices, we need to also query for edges
+      if (vertices.find(v => v.type === "vertex")) {
+        try {
+          var edges = await this.executeQuery(queryId, gremlinQuery + ".bothE()");
+          results = results.concat(edges);
+        } catch (edgesError) {
+          // Swallow and just return vertices
+          console.log("Error querying for edges: ", (edgesError.message || edgesError));
+          // TODO telemetry?
+        }
+      }
+    } catch (error) {
+      // If there's an error, send it to the client to display
+      this._socket.emit("showQueryError", queryId, error.message || error);
```

Comment:
Can we make things like `showQueryError` into constants or enums or something? Seems like it's begging for spelling-mistake bugs

Reply:
@EricJizbaMSFT Sharing files between these two contexts is difficult because I can't import from a non-Node context.  Do you know of a way to share that doesn't need import?

Reply:
@StephenWeatherford Oh didn't realize that. So you can't import inside graphClient.ts - but is there anything stopping you from exporting from that file?

Reply:
Added TODO